README.md
<<<<<<< 0855384d43d2e83c69bb9ff96f3ed7ead8da615a
![SeqPlots](https://github.com/Przemol/seqplots/wiki/img/ico_full_color.png)
=================================================================================
=======
SeqPlots
========
>>>>>>> Adds rain/ TSCAN/ GOsummaries/ geecc/ seqplots/ systemPipeR/ to the repos.

> An interactive tool for visualizing track signals and sequence motif densities along genomic features using average plots and heatmaps

[![Build Status](https://travis-ci.org/Przemol/seqplots.svg?branch=master)](https://travis-ci.org/Przemol/seqplots)
<<<<<<< 0855384d43d2e83c69bb9ff96f3ed7ead8da615a
[![codecov.io](http://codecov.io/github/Przemol/seqplots/coverage.svg?branch=master)](http://codecov.io/github/Przemol/seqplots?branch=master)
[![How long since the package was first in a released Bioconductor version (or is it in devel only)](http://bioconductor.org/shields/years-in-bioc/seqplots.svg)](http://bioconductor.org/packages/seqplots)

![Examples of Seq Plots interface and outputs](https://github.com/Przemol/seqplots/wiki/img/SeqPlots_fig1_web.png)

:exclamation: **RELEASE NOTE** :exclamation:

GitHub repository holds development version of SeqPlots. The release is available on Bioconductor.
Please report the problem, bugs, unexpected behaviors and missing features using [**issue tracker**](../issues).

Live demo
---------

The working demonstration preview of SeqPlots is available on ShinyApps.io:
**https://seqplots.shinyapps.io/seqplots/**

Demo limits:

-	the application contains some pre-loaded data (*C. elegans* genes and histone modifications)
-	size of uploaded files is limited to 30MB, please bin or subset your tracks to fit within the limit (due to ShinyApps beta upload limitations)
-	only *C. elegans* genome is available (due ShinyApps beta space limits)
-	uploaded files will be deleted after some time

Introduction
------------

=======

![Examples of Seq Plots interface and outputs](https://github.com/Przemol/seqplots/wiki/img/00_IntroS.png)

## Introduction
>>>>>>> Adds rain/ TSCAN/ GOsummaries/ geecc/ seqplots/ systemPipeR/ to the repos.
The SeqPlots is universal, web browser based tool for plotting
average track signal (e.g. reads coverage) and sequence motif density over
user specified genomic features. The data are visualized on linear plot,
with error estimates as fields, or as series of heatmaps that can be sorted
and clustered. The dual-purpose implementation allows running the software
locally on desktop or deploying it on server. The unique features of our
software are collaboration and data sharing capabilities, as well as
ability to store pre-calculated result matrixes, that combine many
sequencing experiments and in-silico generated tracks with multiple
different features. These binaries can be further used to generate new
combination plots on fly, run automated batch operations or share with
colleagues, who can adjust their plotting parameters without loading actual
tracks and recalculating numeric values.

<<<<<<< 0855384d43d2e83c69bb9ff96f3ed7ead8da615a
Installation
------------

Install develpment version from Bioconductor:

```{r}
source("http://bioconductor.org/biocLite.R")
biocLite("seqplots")
```

To install GitHub vesrion (in order to solve dependencies install from Bioconductor first):
```{r}
if (!require("devtools")) install.packages("devtools")
devtools::install_github('przemol/seqplots', build_vignettes=FALSE)
```

Quick start
-----------

To start web browser GUI:

```{r}
library(seqplots)
run()
```

To start with R scripting mode:

```{r}
?getPlotSetArry
```

Mac OS X bundle app
-------------------
Mac OS X bundle is an easy way to use SeqPlots for Mac OS X users. It contains R binaries pre-installed Bioconductor dependencies. Additionally, the reference genome for *Caenorhabditis elegans* is included. The sequences for other popular model organisms can be downloaded using graphical user interface.

**Download the current version here:
https://github.com/Przemol/seqplots_osx/releases/latest**

The bundle requires OS X 10.6 (Snow Leopard) and above with X11 installed. X11 was included with OS X up to version 10.7 (Lion). OS X 10.8 (Mountain Lion) and above require installation of Xquartz package, see more here: http://support.apple.com/en-us/HT201341

Docker image
------------
SeqPlots is available as Docker image, which can be easily run on Mac OS, Windows and Linux systems. The image can be obtained from Docker Hub: https://hub.docker.com/r/przemol/seqplots/

Refer to https://www.docker.com/ for instructions how to install Docker and run the image.

References
----------

**R project and Bioconductor**

-   R Core Team (2014). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. URL http://www.R-project.org/.
-   Bioconductor: Open software development for computational biology and bioinformatics R. Gentleman, V. J. Carey, D. M. Bates, B.Bolstad, M.Dettling, S. Dudoit, B. Ellis, L. Gautier, Y. Ge, and others 2004, Genome Biology, Vol. 5, R80.  URL http://www.bioconductor.org/.
- RStudio and Inc. (2014). shiny: Web Application Framework for R. R package version 0.10.1. http://shiny.rstudio.com/
- **Other CRAN packages:** digest, DBI,RSQLite, RJSONIO, plotrix, fields, grid, kohonen, Cairo and parallel
- **Bioconductor packages:** IRanges, BSgenome, Rsamtools, rtracklayer, GenomicRanges and Biostrings 

**JavaScript and CSS**

- jQuery framework - http://jquery.com
- Bootstrap - http://getbootstrap.com
- DataTables, Table plug-in for jQuery - http://www.datatables.net
- jQuery File Upload Plugin - https://github.com/blueimp/jQuery-File-Upload
- jQuery throttle - http://benalman.com/projects/jquery-throttle-debounce-plugin/
- jQuery Cookie Plugin - https://github.com/carhartl/jquery-cookie
- Modernizer JS library - http://modernizr.com
- JavaScript Templates - https://github.com/blueimp/JavaScript-Templates
- JavaScript Color Picker - http://jscolor.com
- md5-js - https://github.com/wbond/md5-js
- Font Awesome - http://fortawesome.github.io/Font-Awesome
- Google Fonts - https://www.google.com/fonts
- jQuery user interface - http://jqueryui.com (documentation)
- jquery.tocify.js: jQuery Table of Contents - https://github.com/gfranko/jquery.tocify.js (documentation)
- Strapdown https://github.com/arturadib/strapdown (documentation)
- Bootswatch themes - http://bootswatch.com (documentation)
- google-code-prettify - https://code.google.com/p/google-code-prettify (documentation)
- marked - https://github.com/chjj/marked (documentation)

**Important conceptual contribution to the project**

- Liu T, Ortiz J, Taing L, Meyer C, Lee B, Zhang Y, Shin H, Wong S, Ma J, Lei Y, et al. 2011. [Cistrome: an integrative platform for transcriptional regulation studies.](http://www.ncbi.nlm.nih.gov/pubmed/21859476) Genome Biology 12: R83.
- Thomas Williams, Colin Kelley and others (2010). Gnuplot 4.4: an interactive plotting program. URL http://www.R-project.org/.
- Kent, W.J., Sugnet, C.W., Furey, T.S., Roskin, K.M., Pringle, T.H., Zahler, A.M. and Haussler, a. D. (2002). [The Human Genome Browser at UCSC](http://www.ncbi.nlm.nih.gov/pubmed/12045153). Genome Research. 12:996–1006.
- Kent WJ, Zweig AS, Barber G, Hinrichs AS, Karolchik D. (2010). [BigWig and BigBed: enabling browsing of large distributed datasets.](http://www.ncbi.nlm.nih.gov/pubmed/20639541) Bioinformatics. 1;26(17):2204-7
- Nicol, J.W., Helt, G.A., Blanchard, S.G., Raja, A. and Loraine, A.E. (2009). [The Integrated Genome Browser: free software for distribution and exploration of genome-scale datasets.](http://www.ncbi.nlm.nih.gov/pubmed/19654113) Bioinformatics (Oxford, England). 25:2730–1.
- Thorvaldsdóttir, H., Robinson, J.T. and Mesirov, J.P. (2012). [Integrative Genomics Viewer (IGV): high-performance genomics data visualization and exploration.](http://www.ncbi.nlm.nih.gov/pubmed/22517427) Briefings in bioinformatics. bbs017


**Server deployment**

- Shiny Server - https://github.com/rstudio/shiny-server
- ShinyApps - https://github.com/rstudio/shinyapps


**Publications containing figures made by SeqPlots**

- Chen RA, Stempor P, Down TA, Zeiser E, Feuer SK, Ahringer J. [Extreme HOT regions are CpG-dense promoters in C. elegans and humans.](http://www.ncbi.nlm.nih.gov/pubmed/24653213) Genome Res 24(7):1138-1146 Jul 2014
- Latorre I, Chesney MA, Garrigues JM, Stempor P et al. [The DREAM complex promotes gene body H2A.Z for target repression](http://www.ncbi.nlm.nih.gov/pubmed/25737279). Genes Dev 2015 Mar 1;29(5):495-500.
=======
## Installation
```{r}
source("http://bioconductor.org/biocLite.R")
biocLite(c("methods", "IRanges", "BSgenome", "digest", "rtracklayer", "GenomicRanges", 
  "Biostrings", "shiny", "DBI", "RSQLite", "RJSONIO", "plotrix", "fields", "grid", 
  "kohonen", "Cairo", "parallel", "GenomeInfoDb", "class", "testthat", "BiocStyle",
  "knitr"))

if (!require("devtools")) install.packages("devtools")
devtools::install_github('przemol/seqplots', build_vignettes=FALSE)
```
>>>>>>> Adds rain/ TSCAN/ GOsummaries/ geecc/ seqplots/ systemPipeR/ to the repos.


NEWS
<<<<<<< 0855384d43d2e83c69bb9ff96f3ed7ead8da615a
CHANGES IN VERSION 1.4.1
--------------------------------------------------------------------------------

PACKAGE:

* RJSONIO replaced with jsonlite package for encoding/decoding JSON
* Sub-heatmaps accepts multilevel color palettes in plotHeatmap function
* Better dendrogram positioning for heatmaps with hclust and single sub-heatmap

GUI:

* "Load saved plot set" select input replaced with searchable selectize input
* Tracks and feature names are shown without extensions
* Plot Preview GUI elements are shown directly after calculation finishes or 
  save dataset si loaded, even if no pair is selected. 
* Download buttons moved to bottom of side panel 
* Color palettes from RColorBrewer package can be selected in GUI
* Web interface tab for managing the reference genomes

BUGFIX:

* Plot Preview shows up when needed with shiny 0.12.0 (issue #6)
* Heatmap top labels and color key font sizing works properly
* Clustering report download works when source feature file were deleted


CHANGES IN VERSION 1.3.0
--------------------------------------------------------------------------------

* New package version for Bioconductor 3.1 release


CHANGES IN VERSION 1.2.0
--------------------------------------------------------------------------------

PACKAGE:

* plot a dendrogram as first panel on heatmap plot clustered with hierarchical 
  clustering
* heatmap can be plotted as vector graphics (default for R scripting) or 
  raster graphics (default for web GUI)
* heatmaps can be sorted in increasing or decreasing order
* y-axis is annotated with cluster IDs and number of data rows (heatmap)
* x-axis is annotated with base pairs, e.g. 10bp, 1kb, 1.2Mb, etc.
* heatmaps can be plotted using GGplot2 package instead of R base graphics 
  system
* speed-up in motif plots thanks to faster, vectored implementation 
  of motif density data acquisition 
* code simplification - seqplots in Shiny GUI mode use all functions 
  from package core

GUI:

* feature-track pairs selection grid have UI elements to multi-select and batch 
  change labels/colors/etc.
* option to make a clustering repeatable (works by reusing a .Random.seed)
* option to plot selected cluster only as heatmap
* SeqPlots GUI updated to work with Bootstrap v3.3.1

BUGFIX:

* lines separating the clusters are drawn in correct places on 
  grDevices::quartz devices
* proper positioning of 3' ends of anchored region in motif plot
* fixed error with anchored plots and genomic feature width equals 1bp
* y-axis does respect cex.axis parameter
* GUI fixed to work with shiny 0.11.0 and above


CHANGES IN VERSION 1.0.0
--------------------------------------------------------------------------------

PACKAGE:

* plotHeatmap function returns cluster report as GRanges structure
* redundant parameters removed from plotting functions
* plotHeatmap function have "embed" parameter for plotHeatmap - allows to 
  plot 1st heatmap without using grid system, intended to use with complex plots

BUGFIX:

* motif plot orientation properly dependents on strand
* GUI - reordering the heatmap respects previously set include/exclude 
  parameters


CHANGES IN VERSION 0.99.1
--------------------------------------------------------------------------------

PACKAGE:

=======
SeqPlots v1.1.0 (next devel)
* TODO: vignette for console (R scripting) mode workflow
* TODO: cluster report as GRanges structure


SeqPlots v0.99.1
--------------------------------------------------------------------------------
PACKAGE:
>>>>>>> Adds rain/ TSCAN/ GOsummaries/ geecc/ seqplots/ systemPipeR/ to the repos.
* heatmap plotting function returns cluster report ad data.table
* getPlotSetArray function have "verbose" parameter that controls messages 
  and warnings output
* references added to documentation


BUGFIX:
<<<<<<< 0855384d43d2e83c69bb9ff96f3ed7ead8da615a

=======
>>>>>>> Adds rain/ TSCAN/ GOsummaries/ geecc/ seqplots/ systemPipeR/ to the repos.
* plotHeatmap and plotAverage generic methods for SeqPlots-classes respect the
  parameters
* package passes tests and check on 32bit Windows (plotting only, because
  no rtrackalyer::BigWigFile support for Win32)


<<<<<<< 0855384d43d2e83c69bb9ff96f3ed7ead8da615a
CHANGES IN VERSION 0.99
--------------------------------------------------------------------------------

GENERAL:

* Anchored plots and heatmaps uses [downstream]--0--0--[upstream] X-axis 
  coordinate system instead [downstream]--0--[anchored]-[upstream+anchored]

PACKAGE:

=======
SeqPlots v0.99
--------------------------------------------------------------------------------
GENERAL:
* Anchored plots and heatmaps uses [downstream]--0--0--[upstream] X-axis coordinate 
  system instead [downstream]--0--[anchored]-[upstream+anchored]

PACKAGE:
>>>>>>> Adds rain/ TSCAN/ GOsummaries/ geecc/ seqplots/ systemPipeR/ to the repos.
* package really on reference class system including MotifSetup, PlotSetArray, 
  PlotSetList and PlotSetPair
* generic subset and data manipulation methods for SeqPlots-classes including 
  '[', "[[" and "unlist", which allows to switch between classes
* automatic tests for class system, calculations and plotting functions
* documentation for all functions and classes
* PDF vignette engine replaced by HTML one
* QuickStart vignette added

GUI:
<<<<<<< 0855384d43d2e83c69bb9ff96f3ed7ead8da615a

* automated GUI tests using Rselenium package

BUGFIX:

* issue #1: some server instances loads empty .Rdata file on startup


CHANGES IN VERSION 0.9.3
--------------------------------------------------------------------------------

=======
* automated GUI tests using Rselenium package

BUGFIX:
* issue #1: some server instances loads empty .Rdata file on startup


SeqPlots v0.9.3-merge
--------------------------------------------------------------------------------
>>>>>>> Adds rain/ TSCAN/ GOsummaries/ geecc/ seqplots/ systemPipeR/ to the repos.

R/seqplots-package.R
#' @docType package
#' @name seqplots
#' 
#' @title 
#' SeqPlots - An interactive tool for visualizing NGS signals and sequence 
#' motif densities along genomic features using average plots and heatmaps.
#' 
#' @description
#' SeqPlots is a tool for plotting next generation sequencing (NGS) based 
#' experiments' signal tracks, e.g. reads coverage from ChIP-seq, RNA-seq 
#' and DNA accessibility assays like DNase-seq and MNase-seq, over user 
#' specified genomic features, e.g. promoters, gene bodies, etc. It can 
#' also calculate sequence motif density profiles from reference genome. 
#' The data are visualized as average signal profile plot, with error estimates 
#' (standard error and 95% confidence interval) shown as fields, or as series 
#' of heatmaps that can be sorted and clustered using hierarchical clustering, 
#' k-means algorithm and self organising maps. Plots can be prepared using R 
#' programming language or web browser based graphical user interface (GUI) 
#' implemented using Shiny framework. The dual-purpose implementation allows 
#' running the software locally on desktop or deploying it on server. 
#' SeqPlots is useful for both for exploratory data analyses and preparing 
#' replicable, publication quality plots. Other features of the software 
#' include collaboration and data sharing capabilities, as well as ability 
#' to store pre-calculated result matrixes, that combine many sequencing 
#' experiments and in-silico generated tracks with multiple different features. 
#' These binaries can be further used to generate new combination plots on fly, 
#' run automated batch operations or share with colleagues, who can adjust 
#' their plotting parameters without loading actual tracks and recalculating 
#' numeric values. SeqPlots relays on Bioconductor packages, mainly on 
#' rtracklayer for data input and BSgenome packages for reference genome 
#' sequence and annotations.
#' 
#' @details
#' Useful links:
#' \itemize{
#'   \item Project web page: \url{https://github.com/Przemol/seqplots} 
#'   \item Online documentation: \url{http://przemol.github.io/seqplots/} 
#'   \item Project wiki: \url{https://github.com/przemol/seqplots/wiki} 
#'   \item Issue tracker: \url{http://github.com/przemol/seqplots/issues} 
#'   \item Bug reports: \url{http://github.com/przemol/seqplots/issues/new}
#' }
#' 
#' @references 
#' \strong{R project and Bioconductor:}
#' \itemize{
#'   \item R Core Team (2014). R: A language and environment for statistical
#'    computing. R Foundation for Statistical Computing, Vienna, Austria. URL
#'    \url{http://www.R-project.org/}. 
#'   \item Bioconductor: Open software development for
#'    computational biology and bioinformatics R. Gentleman, V. J. Carey, D. M.
#'    Bates, B.Bolstad, M.Dettling, S. Dudoit, B. Ellis, L. Gautier, Y. Ge, and
#'    others 2004, Genome Biology, Vol. 5, R80.  URL
#'    \url{http://www.bioconductor.org/}. 
#'   \item RStudio and Inc. (2014). shiny: Web
#'    Application Framework for R. R package version 0.10.1.
#'    \url{http://shiny.rstudio.com/}.
#'   \item Other CRAN packages: digest,
<<<<<<< 0855384d43d2e83c69bb9ff96f3ed7ead8da615a
#'    DBI,RSQLite, jsonlite, plotrix, fields, grid, kohonen, Cairo, parallel 
=======
#'    DBI,RSQLite, RJSONIO, plotrix, fields, grid, kohonen, Cairo, parallel 
>>>>>>> Adds rain/ TSCAN/ GOsummaries/ geecc/ seqplots/ systemPipeR/ to the repos.
#'   \item Bioconductor packages: IRanges, BSgenome, Rsamtools, rtracklayer,
#'    GenomicRanges and Biostrings
#'   }
#'   
#' \strong{JavaScript and CSS}
#' \itemize{
#'   \item jQuery framework - \url{http://jquery.com}
#'   \item Bootstrap - \url{http://getbootstrap.com}
#'   \item DataTables, Table plug-in for jQuery - 
#'    \url{http://www.datatables.net}
#'   \item jQuery File Upload Plugin - 
#'    \url{https://github.com/blueimp/jQuery-File-Upload}
#'   \item jQuery throttle - 
#'    \url{http://benalman.com/projects/jquery-throttle-debounce-plugin/}
#'   \item jQuery Cookie Plugin - 
#'    \url{https://github.com/carhartl/jquery-cookie}
#'   \item Modernizer JS library - \url{http://modernizr.com}
#'   \item JavaScript Templates - 
#'    \url{https://github.com/blueimp/JavaScript-Templates}
#'   \item JavaScript Color Picker - \url{http://jscolor.com}
#'   \item md5-js - \url{https://github.com/wbond/md5-js}
#'   \item Font Awesome - \url{http://fortawesome.github.io/Font-Awesome}
#'   \item Google Fonts - \url{https://www.google.com/fonts}
#'   \item jQuery user interface - \url{http://jqueryui.com} (documentation)
#'   \item jquery.tocify.js: jQuery Table of Contents - 
#'    \url{https://github.com/gfranko/jquery.tocify.js} (documentation)
#'   \item Strapdown \url{https://github.com/arturadib/strapdown} 
#'    (documentation)
#'   \item Bootswatch themes - \url{http://bootswatch.com} (documentation)
#'   \item google-code-prettify - 
#'    \url{https://code.google.com/p/google-code-prettify} (documentation)
#'   \item marked - \url{https://github.com/chjj/marked} (documentation)
#' }
#' 
#' \strong{Important conceptual contribution to the project}
#' \itemize{
#' \item Liu T, Ortiz J, Taing L, Meyer C, Lee B, Zhang Y, Shin H, Wong S, Ma J,
#' Lei Y, et al. 2011. 
#' \href{http://www.ncbi.nlm.nih.gov/pubmed/21859476}{Cistrome: an integrative 
#' platform for transcriptional regulation studies.} Genome Biology 12: R83. 
#' \item Thomas Williams, Colin Kelley and others (2010).
#' Gnuplot 4.4: an interactive plotting program. URL 
#' \url{http://www.R-project.org/}. 
#' \item Kent, W.J., Sugnet, C.W., Furey, T.S., Roskin, K.M., Pringle, T.H.,
#' Zahler, A.M. and Haussler, a. D. (2002). 
#' \href{http://www.ncbi.nlm.nih.gov/pubmed/12045153}{The Human Genome Browser 
#' at UCSC}. Genome Research. 12:996-1006. 
#' \item Kent WJ, Zweig AS, Barber G, Hinrichs AS, Karolchik D.
#' (2010). \href{http://www.ncbi.nlm.nih.gov/pubmed/20639541}{BigWig and BigBed:
#'  enabling browsing of large distributed datasets.} Bioinformatics.
#' 1;26(17):2204-7 
#' \item Nicol, J.W., Helt, G.A., Blanchard, S.G., Raja, A. and Loraine, A.E. 
#' (2009). \href{http://www.ncbi.nlm.nih.gov/pubmed/19654113}{The Integrated 
#' Genome Browser: free software for distribution and exploration of 
#' genome-scale datasets.} Bioinformatics (Oxford, England). 25:2730-1. 
#' \item Thorvaldsdottir, H., Robinson, J.T. and Mesirov, J.P. (2012). 
#' \href{http://www.ncbi.nlm.nih.gov/pubmed/22517427}{Integrative Genomics 
#' Viewer (IGV): high-performance genomics data visualization and
#' exploration.} Briefings in bioinformatics. bbs017 
#' }
#' 
#' \strong{Server deployment}
#' \itemize{
#'   \item Shiny Server - \url{https://github.com/rstudio/shiny-server}
#'   \item ShinyApps - \url{https://github.com/rstudio/shinyapps}
#' }
#' 
#' \strong{Publications containing figures made by SeqPlots}
#' \itemize{
#'   \item Chen RA, Stempor P, Down TA, Zeiser E, Feuer SK, Ahringer J. 
#'   \href{http://www.ncbi.nlm.nih.gov/pubmed/24653213}{Extreme HOT regions are 
#'   CpG-dense promoters in C. elegans and humans.} Genome Res 
#'   24(7):1138-1146 Jul 2014
#'} 
#' 
#' 
#' @author Przemyslaw Stempor
#' 
#' @import rtracklayer GenomicRanges BSgenome Biostrings IRanges GenomeInfoDb
<<<<<<< 0855384d43d2e83c69bb9ff96f3ed7ead8da615a
#' @import DBI RSQLite parallel Cairo jsonlite
#' @import grid methods digest methods
#' @import ggplot2 reshape2 gridExtra DT RColorBrewer
=======
#' @import DBI RSQLite parallel RJSONIO Cairo
#' @import grid methods digest methods
>>>>>>> Adds rain/ TSCAN/ GOsummaries/ geecc/ seqplots/ systemPipeR/ to the repos.
#' @importFrom fields image.plot set.panel imageplot.info imageplot.setup 
#' tim.colors poly.image
#' @importFrom kohonen supersom
#' @importFrom plotrix dispersion
#' @importFrom shiny runApp
#' @importFrom class somgrid
#' @importFrom S4Vectors recycleIntegerArg
NULL
